### PR TITLE
[FEATURE] Client Credentials authentication flow

### PR DESCRIPTION
**Pull request by celian-garcia**
# Description

Implement the client credentials flow. It will allow to authenticate in Perses through CLI or API using a robotic access like it would be needed in scripted pipeline for example. 

A specification documentation with a sequence diagram is coming, but the logic is the same as the auth code/device code flow, we send a query to /providers/{oidc|oauth}/{slug_id}/token, it asks a token to the provider and generate a Perses access token from it. 

It performs also a user synchronization to save a user using a ``metadata.name`` as good as we can. The decision of what to take as ``metadata.name`` is not easy but the ``GetLogin`` of both oauth and oidc are well encapsulated and quite explicit. We can improve but for now it is good enough in my opinion.

image

# Screenshots

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the
  following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `DOC`,`IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

### CODE REVIEW

File: internal/api/impl/auth/oidc.go [hunk 1]
```diff
@@ -367,8 +371,10 @@ func retrieveDeviceAccessToken(ctx context.Context, relyingParty RelyingPartyWit
 func (e *oIDCEndpoint) retrieveClientCredentialsToken(ctx context.Context, clientID string, clientSecret string) (*oauth2.Token, error) {
 	// Create a new client credentials request
 	req := &oidc.ClientCredentialsRequest{
+		GrantType:    oidc.GrantTypeClientCredentials,
 		ClientID:     clientID,
 		ClientSecret: clientSecret,
+		Scope:        e.relyingParty.OAuthConfig().Scopes,
```
Comment (celian-garcia) [inline internal/api/impl/auth/oidc.go:377]: TODO in azure, the scope is different for client credentials flow. This needs to be configurable.

File: internal/cli/cmd/login/login.go [hunk 2]
```diff
@@ -192,28 +210,62 @@ func (o *option) readAndSetLoginInputNative() error {
...
 	for _, prov := range providers.OIDC {
 		if prov.SlugID == slugID {
-			return o.setLoginInputExternal(externalAuthKindOIDC, slugID)()
+			return o.setLoginInputExternal(externalAuthKindOIDC, slugID)
 		}
 	}
 	for _, prov := range providers.OAuth {
 		if prov.SlugID == slugID {
-			return o.setLoginInputExternal(externalAuthKindOAuth, slugID)()
+			return o.setLoginInputExternal(externalAuthKindOIDC, slugID)
```
Comment (celian-garcia) [inline internal/cli/cmd/login/login.go:245]: nope

File: pkg/client/api/auth.go [hunk 3]
```diff
@@ -26,10 +26,11 @@ const authResource = "auth"
 
 // AuthInterface has methods to work with Auth resource
 type AuthInterface interface {
-	DeviceCode(authKind, authProvider string) (*oauth2.DeviceAuthResponse, error)
-	DeviceAccessToken(authKind, slugID, deviceCode string) (*api.AuthResponse, error)
 	Login(user, password string) (*api.AuthResponse, error)
 	Refresh(refreshToken string) (*api.AuthResponse, error)
+	DeviceCode(authKind, authProvider string) (*oauth2.DeviceAuthResponse, error)
+	DeviceAccessToken(authKind, slugID, deviceCode string) (*api.AuthResponse, error)
+	Token(authKind, slugID, clientID, clientSecret string) (*api.AuthResponse, error)
```
Comment (celian-garcia) [inline pkg/client/api/auth.go:33]: Rename ClientCredsToken

File: internal/api/impl/auth/oauth.go [hunk 4]
```diff
@@ -66,7 +106,11 @@ func (u *oauthUserInfo) GetLogin() string {
 	if login := u.getProperty(u.loginKeys); login != "" {
 		return login
 	}
-	return buildLoginFromEmail(u.Email)
+	if login := buildLoginFromEmail(u.Email); login != "" {
+		return login
+	}
+	//TODO: Find something to return if we don't find anything to us as a login. Maybe an error ?
+	return "TODO"
```
Comment (celian-garcia) [inline internal/api/impl/auth/oauth.go:113]: Solve this

File: pkg/model/api/config/authentication.go [hunk 5]
```diff
@@ -45,14 +46,15 @@ func appendIfMissing[T comparable](slice []T, value T) ([]T, bool) {
...
-	DeviceCode   DeviceCode    `json:"device_code,omitempty" yaml:"device_code,omitempty"`
-	RedirectURI  common.URL    `json:"redirect_uri" yaml:"redirect_uri"`
-	Scopes       []string      `json:"scopes" yaml:"scopes"`
-	DisablePKCE  bool          `json:"disable_pkce" yaml:"disable_pkce"`
+	SlugID            string         `json:"slug_id" yaml:"slug_id"`
+	Name              string         `json:"name" yaml:"name"`
+	ClientID          secret.Hidden  `json:"client_id" yaml:"client_id"`
+	ClientSecret      secret.Hidden  `json:"client_secret" yaml:"client_secret"`
+	DeviceCode        *OAuthOverride `json:"device_code,omitempty" yaml:"device_code,omitempty"`
+	ClientCredentials *OAuthOverride `json:"client_credentials,omitempty" yaml:"client_credentials,omitempty"`
```
Comment (celian-garcia) [inline pkg/model/api/config/authentication.go:54]: Don´t  forget to document it

File: internal/api/impl/auth/oauth.go (hunk 4 above)
Comment (celian-garcia) [reply internal/api/impl/auth/oauth.go]: In the end, an empty will be an error

**Review by Nexucis — APPROVED** [review]
🥳

**Pull request merged**